### PR TITLE
Bug 1837362: Sensitive data exposed on pipeline failure

### PR DESCRIPTION
**Pull request by rinormaloku**
Resolves the ticket #326 

i.e. the following output:
```
{err=Error from server (BadRequest): error when creating "/Users/rmaloku/work/jenkins-client-plugin/work/workspace/test-ocp/apply2288494789001734078.markup": Secret in version "v1" cannot be handled as a Secret: v1.Secret.Data: decode base64: illegal base64 data at input byte 4, error found in #10 byte of ...|t":"xxxxK","secretKe|..., bigger context ...|{"apiVersion":"v1","data":{"secret":"xxxxK","secretKey":"YWRtaW4K"},"kind":"Secret","metadata|...
``` 
is printed as:
```
{err=Error from server (BadRequest): Secret in version "v1" cannot be handled as a Secret: v1.Secret.Data: decode base64: illegal base64 data at input byte 4, error found in #10 byte of ...|t":"xxxxK","secretKe|..., bigger context ...|{"apiVersion":"v1","data":{ REDACTED },"kind":"Secret","metadata|...
```

The important section is: **"data":{ REDACTED }**

FYI: I didn't hide the following section `|t":"xxxxK","secretKe|` as it provides the user with information of where the mistake is. Though I am open for your suggestions if we want to hide this as well.

### CODE REVIEW

File: src/main/java/com/openshift/jenkins/plugins/freestyle/BaseStep.java [hunk 1]
```diff
@@ -255,16 +260,18 @@ public boolean perform(ProcessBuilder pb)
                         new Thread(new Runnable() {
                             @Override
                             public void run() {
-                                byte buffer[] = new byte[1024];
-                                int count;
-                                try {
-                                    while ((count = output.read(buffer)) != -1) {
-                                        listener.getLogger().write(buffer, 0,
-                                                count);
+                                StringBuilder sb = new StringBuilder();
```
Comment (akram) [inline src/main/java/com/openshift/jenkins/plugins/freestyle/BaseStep.java:263]: StringBuilder is not synchronized, so, we may endup in log messages displayed in a different order from what they are produced.
Any objections in using StringBuffer instead ?

File: src/main/java/com/openshift/jenkins/plugins/pipeline/OcAction.java [hunk 2]
```diff
@@ -63,14 +64,18 @@ public OcAction(String server, String project, boolean skipTLSVerify, String caP
         public int status;
         @Whitelisted
         public HashMap<String, String> reference = new HashMap<String, String>();
+        @Whitelisted
+        public boolean verbose = false;
```
Comment (akram) [inline src/main/java/com/openshift/jenkins/plugins/pipeline/OcAction.java:68]: should this flag be propagated to all the commands in the GroovyDSL?
if so, we can do it in a separate PR.

Comment (rinormaloku) [reply src/main/java/com/openshift/jenkins/plugins/pipeline/OcAction.java:68]: I forwarded it from the Step verbose property, to ensure that OcActionResult doesn't print the entire yaml/json on Jenkins. 
But I might misunderstand the question and would love suggestions on how to improve it.

File: src/main/java/com/openshift/jenkins/plugins/freestyle/BaseStep.java (hunk 1 above)
Comment (rinormaloku) [reply src/main/java/com/openshift/jenkins/plugins/freestyle/BaseStep.java]: @akram Thanks for the correction, I was not aware of the difference.

**Comment by akram** [comment]
/test e2e-aws-jenkins

looks like an infra flake, static pods cannot listen on port 6443

**Comment by akram** [comment]
/retitle Bug 1837362: Sensitive data exposed on pipeline failure

**Comment by rinormaloku** [comment]
Hi Jitendar,

To test the freestyle I just created one resource:

apiVersion: v1
kind: Secret
metadata:
  name:  secretname
data:
   secretKey: YWRtaW4K
   secret: xxxxK
type: Opaque

This should fail and hide the secret YWRtaW4K i.e. it should not be visible.

testing the pipeline was a little harder, I created an OpenShift template
with the above resource and then applied the template with
openshift.apply().

I'd provide all the scripts, but sadly my jenkins workplace is reset and I
have lost all those.

Best,
Rinor
